gluecodium/src/main/resources/templates/ffi/FfiCallbacksQueueImpl.mustache

```diff
@@ -0,0 +1,170 @@
+{{!!
+  !
+  ! Copyright (C) 2016-2020 HERE Europe B.V.
+  !
+  ! Licensed under the Apache License, Version 2.0 (the "License");
+  ! you may not use this file except in compliance with the License.
+  ! You may obtain a copy of the License at
+  !
+  !     http://www.apache.org/licenses/LICENSE-2.0
+  !
+  ! Unless required by applicable law or agreed to in writing, software
+  ! distributed under the License is distributed on an "AS IS" BASIS,
+  ! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ! See the License for the specific language governing permissions and
+  ! limitations under the License.
+  !
+  ! SPDX-License-Identifier: Apache-2.0
+  ! License-Filename: LICENSE
+  !
+  !}}
+{{>ffi/FfiCopyrightHeader}}
+
+#include "CallbacksQueue.h"
+
+{{#internalNamespace}}
+namespace {{.}}
+{
+{{/internalNamespace}}
+namespace ffi
+{
+CallbackQueueManager cbqm;
+
+CallbackQueue::CallbackQueue() : closed(false) {}
+
+std::future<bool>
+CallbackQueue::enqueueIncoming(std::function<void()> func)
+{
+    std::unique_lock<std::mutex> lock(mutex);
+
+    if (closed) {
+        std::promise<bool> done;
+        done.set_value(false);
+        return done.get_future();
+    }
+
+    if (!incoming) {
+        incoming = true;
+        notified.notify_one();
+    }
+
+    queue.emplace(Entry{std::move(func), std::promise<bool>()});
+
+    return queue.back().done.get_future();
+}
+
+bool
+CallbackQueue::waitForIncoming()
+{
+    std::unique_lock<std::mutex> lock(mutex);
+
+    while (!incoming && !closed) {
```

Comment:
Let the STL do all necessary work, write
`notified.wait(lock, [this]{return m_has_incoming || m_is_closed;});`
instead of loop

Reply:
Done